Add a cancel button to allow merchants to cancel the flow of creating a new order.

Juanpe opened the pull request:
Closes: #6500 


### Description
As the linked issue describes, the flow of creating a new order can't be cancelled/dismissed by using an interactive drag-down on landscape orientation. With this PR, we address this problem.

To fix it, we follow the approach proposed in the issue, adding a new cancel button to allow merchants to cancel/dismiss the flow. This new button has been added to the navigation bar following our system design rules. Also, before dismissing the view, we must validate if there are unsaved changes to ask the merchant before discarding them.

I found out we have similar behaviour on the [Edit Order Address Form](https://github.com/woocommerce/woocommerce-ios/blob/e94ec2ed1db5c51ba26318afd54092c6ca719a71/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Address%20Edit/EditOrderAddressForm.swift#L84) and I've used the same approach, but since I'm not an expert on SwiftUI, please let me know if there is a better way to do it 🙂

In addition, I've added a UI test to cover this behaviour, I mean, the new order screen is dismissed when the cancel button is tapped and there are no unsaved changes.

### Testing instructions

#### Scenario 1️⃣: Cancel the creation of a new order with no unsaved changes

1. Go to Orders and tap the + icon. 
2. Select "Create order".
3. Tap the "Cancel" button.
4. Check that the new order screen is dismissed.

**Demo** 

https://user-images.githubusercontent.com/1409041/161952809-0ea5a0e7-f82b-4b34-9725-4f3edfab78d0.mp4

#### Scenario 2️⃣: Cancel the creation of a new order discarding unsaved changes

1. Go to Orders and tap the + icon. 
2. Select "Create order".
3. Add details to your new order: order status, products, shipping, etc.
4. Tap the "Cancel" button.
5. Select "Discard changes" in the action sheet. 
6. Check that the new orders screen is dismissed and no new orders appear.

**Demo**

https://user-images.githubusercontent.com/1409041/161952790-d0d18ff4-74a1-4dc4-ab28-0171156b1a7c.mp4

#### Scenario 3️⃣: No cancel the creation of a new order with unsaved changes

1. Go to Orders and tap the + icon. 
2. Select "Create order".
3. Add details to your new order: order status, products, shipping, etc.
4. Tap the "Cancel" button.
5. Select "Cancel" in the action sheet. 
6. Check that the new order screen is NOT dismissed.

**Demo**

https://user-images.githubusercontent.com/1409041/161952812-f729fb04-4cda-4791-bbf7-ba9edfb4d0dc.mp4

### Screenshots

|Portrait|Landscape|
|--|--|
|<img width="250px" src="https://user-images.githubusercontent.com/1409041/161952947-458d51d9-0308-42af-aec7-b3ad66f1a8b7.png"/>|<img width="450px" src="https://user-images.githubusercontent.com/1409041/161952941-3040aade-56c4-4ffa-b243-6444591d8b1a.png"/>|


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Juanpe committed:
Add “Cancel” button to discard the creation of an order

Juanpe committed:
Add UITest for cancellation of a new order

Juanpe committed:
Merge branch 'trunk' of https://github.com/Juanpe/woocommerce-ios into issue/6500-add-button-to-dismiss-new-order-modal

# Conflicts:
#	WooCommerce/UITestsFoundation/Screens/Orders/NewOrderScreen.swift

Juanpe committed:
Add fix release note

Juanpe committed:
Remove comment manually added to the Localizable.strings files

Juanpe committed:
Remove missing changes on Localizable.strings files

Juanpe committed:
Revert wrong changes on the English Localizable string file

Juanpe committed:
Revert the last two wrong changes on the English localizable strings file.

Juanpe commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -49,13 +54,6 @@ extension NewOrderHostingController {
         return shouldPopOnBackButton()
     }
-
-    private func presentDiscardChangesActionSheet() {
```
After discarding the changes we can either dismiss or pop the view, so to avoid duplicating the code I've created a method that presents the action sheet and received the `onDiscard` closure as a param.

itsmeichigo replied:
Good thinking about avoiding duplicated code! However, as mentioned above, we only have two ways of closing this flow due to the handling of the feature flag for the split view. We will need to remove the code for popping the view after the feature flag is enabled. Should we revert the changes for simplicity?

Juanpe replied:
Good question! TBH although we only have two ways of closing this flow, we have three "triggers"; the back button (pop), the cancel button (dismiss) and the swipe down (dismiss), so we could leave them to not to do the same two times (cancel and swipe). But it's true that we could have only one method for both triggers, something like `presentDiscardChangesActionSheetAndDismissOnDiscard` 🤔 You know what I mean?

That said, I don't have a strong preference on this, so I might as well revert the changes 🙂

itsmeichigo replied:
> so we could leave them to not to do the same two times (cancel and swipe)

This makes sense 👍 Let's keep the changes then!

Juanpe commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -213,11 +245,17 @@ private extension NewOrder {
                                                                    comment: "Accessibility hint for selecting a product in a new order")
     }
+
+    enum Accessibility {
```
In order to group all the accessibility identifiers, I've created this `enum`, as we do with the localizable and the layout constants. WDYT? 🤔

itsmeichigo replied:
This looks good to me! 👍

Juanpe commented on RELEASE-NOTES.txt:
```diff
@@ -4,6 +4,7 @@
 -----
 
 - [internal] Reviews lists on Products and Menu tabs are refactored to avoid duplicated code. Please quickly smoke test them to make sure that everything still works as before. [https://github.com/woocommerce/woocommerce-ios/pull/6553]
+- [*] Fixed issue that did not allow to merchants cancel the flow of creating a new order on landscape orientation. [https://github.com/woocommerce/woocommerce-ios/pull/6578]
```
I'm not sure if I used the correct priority indicator 😅 

```
[**] → Changes our users will probably notice, but doesn’t impact core flows. Most fixes.
[*] → Minor enhancements and fixes that address annoyances — things our users can miss.
```

Considering the doc, the priority indicator might as well be two asterisks. WDYT? 🤔

itsmeichigo replied:
Good question! In my (very subjective) opinion,  just one asterisk should work since most users probably use portrait mode for the app.

However, this release note update shouldn't be necessary since our Order Creation flow is now only displayed as a modal under the Split View feature flag. In other words, users aren't seeing this issue in the production build at the moment, so this is not a user-facing update.

Juanpe replied:
I got it 👍🏼 Thanks for the clarification 🙂

Juanpe committed:
Fix potential memory leaks

Juanpe committed:
Update UI Tests doc

Juanpe committed:
Merge branch 'trunk' into issue/6500-add-button-to-dismiss-new-order-modal

Juanpe committed:
No escape onDiscard closure

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -113,14 +139,20 @@ struct NewOrder: View {
+                Button(Localization.cancelButton) {
+                    dismiss()
+                }
+                .accessibilityIdentifier(NewOrder.Accessibility.cancelButtonIdentifier)
```
Super nit: you can access `Accessibility` directly since this code is called from within `NewOrder`.
```suggestion
.accessibilityIdentifier(Accessibility.cancelButtonIdentifier)
```

Juanpe replied:
yep, much better 🙂

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -113,14 +139,20 @@ struct NewOrder: View {
                     }
                     .id(navigationButtonID)
-                    .accessibilityIdentifier("new-order-create-button")
+                    .accessibilityIdentifier(NewOrder.Accessibility.createButtonIdentifier)
```
Same comment as above about accessing `Accessibility` directly:
```suggestion
                    .accessibilityIdentifier(Accessibility.createButtonIdentifier)
```

Juanpe replied:
👍🏼

itsmeichigo commented on docs/UI-TESTS.md:
```diff
@@ -18,6 +18,7 @@ The following flows are covered/planned to be covered by UI tests. Tests that ar
     - [ ] Tap chart on stats
 3. [Orders](../WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift)
     - [x] Orders list and single order screens load
+    - [x] Cancel the flow of creation of a new order
```
Nit: This can be clarified that the test checks only the canceling of a fresh order (without any input).
```suggestion
    - [x] Dismissal of order creation flow without any changes made
```

Juanpe replied:
Yep, it makes sense. Thanks for the suggestion :)

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -66,16 +64,44 @@ extension NewOrderHostingController: UIAdaptivePresentationControllerDelegate {
+        self.dismiss(animated: true)
+    }
+
+    func presentDiscardChangesActionSheet(onDiscard: (() -> Void)?) {
```
❓ I'm curious - could you explain why we need to keep the `onDiscard` closure optional?

Juanpe replied:
It's a good question 😀 The question for me was why we need to escape this closure if our `UIAlertController`doesn't do it 🤔 And I couldn’t find any good reason. However, after knowing the feature flag, the rules changes, so I think this method will disappear 🗑️

itsmeichigo replied:
> The question for me was why we need to escape this closure if our UIAlertControllerdoesn't do it

I just realized that the `onDiscard` closure of the `UIAlertController` extension method is optional too, that is why it's escaping. It's weird that the closure needs to be optional since we're more likely to trigger some action when discarding changes 🤷‍♀️ .

Juanpe replied:
Yep, it's weird 😅 In my opinion, if the view is destroyed there is no point in keeping the closures, however, I don't see any downside if we do 😅 So what do you think, should we escape them?

itsmeichigo replied:
I believe the closure needs to be escaped if it's not optional, since it will be called as part of a `UIAlertAction`. So it's not about whether it is necessary to keep the closure, WDYT?

Juanpe replied:
ok, it's a good point, I'm going to make the change 👍🏼 
Thanks! 🙂

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -66,16 +64,44 @@ extension NewOrderHostingController: UIAdaptivePresentationControllerDelegate {
 struct NewOrder: View {
+    /// Set this closure with UIKit dismiss code. Needed because we need access to the UIHostingController `dismiss` method.
+    ///
+    var dismiss: (() -> Void) = {}
```
Nit: since this is a variable, how about renaming it to `dismissHandler` instead? `dismiss` is a verb and I think it's more suitable for a method 😅

Juanpe replied:
Totally agree with you 🙂 In fact, I was thinking of changing the name to use something like `cancelButtonTapHandler`, but then, I thought that this closure is because this view is embedded in a `UIHostingController`, so I think we should still use `dismissHandler` to be clear what the closure should do. 
WDYT?

PS: Naming problems, I love them ❤️ (no irony 🙂)

itsmeichigo replied:
Yes, `dismissHandler` seems to be the best candidate in this case.

itsmeichigo commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -32,4 +32,10 @@ final class OrdersTests: XCTestCase {
             .createOrder()
     }
+
+    func test_cancel_order_creation() throws {
+        try TabNavComponent().goToOrdersScreen()
+            .startOrderCreation()
+            .cancelOrderCreation()
+    }
```
Thank you for adding the UI test!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -113,14 +139,20 @@ struct NewOrder: View {
         .navigationTitle(Localization.title)
         .navigationBarTitleDisplayMode(.inline)
         .toolbar {
+            ToolbarItem(placement: .cancellationAction) {
```
This item should be added only when the `.splitViewInOrdersTab` feature flag is enabled.

itsmeichigo reviewed (changes requested):
Hi Juanpe - I tested the changes and found a case that doesn't work as expected. As mentioned in #6500:

> The Order Creation flow gets displayed as a modal under the Split View feature flag.

The backstory of this is that we are working on #6381 to add split view to the Orders tab. Previously the Order Creation flow was pushed to the navigation stack directly - but we had to update this flow to present the New Order screen as a modal to make it look good when the split view is introduced. The changes are now hidden under [a feature flag](https://github.com/woocommerce/woocommerce-ios/blob/667504644115ef4b635952a20d2f17b888eeb805/Experiments/Experiments/FeatureFlag.swift#L55) until the feature is finished.

You can test how the app currently looks like in production by [disabling](https://github.com/woocommerce/woocommerce-ios/blob/667504644115ef4b635952a20d2f17b888eeb805/Experiments/Experiments/DefaultFeatureFlagService.swift#L29) the `.splitViewInOrdersTab` feature flag.

We'd expect to only display the Cancel button when the New Order view is presented modally - otherwise, we'll have an additional button next to the back button like this:

<img src="https://user-images.githubusercontent.com/5533851/162123862-00e0d916-0e46-4536-8005-69568b1a7694.png" width=320 />

Suggestion: you can check for the feature flag using `ServiceLocater.featureflagService.isFeatureFlagEnabled(.splitViewInOrdersTab)`.

Juanpe committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift

Remove unnecessary use of class name to access the `Accessibility` namespace for the cancel button

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

Juanpe committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift

Remove unnecessary use of class name to access the `Accessibility` namespace for the create button

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

Juanpe committed:
Update docs/UI-TESTS.md

Clarify what the UI test checks in the UI Tests doc.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

Juanpe commented:
> > The Order Creation flow gets displayed as a modal under the Split View feature flag.

Oups 🤦🏼‍♂️ my fault, I didn't realize we are using a feature flag here. I'll modify the code 🙂

Thanks @itsmeichigo for the review and the feedback! 🤙

Juanpe committed:
Merge branch 'trunk' into issue/6500-add-button-to-dismiss-new-order-modal

Juanpe committed:
Remove fix from release notes

Juanpe committed:
Add a condition to render the cancel button only when the Split View feature flag is enable

Juanpe committed:
Rename `dismiss` closure to `dismissHandler`

Juanpe committed:
Move feature flag checking logic to the view model

Juanpe committed:
Merge remote-tracking branch 'origin/trunk' into issue/6500-add-button-to-dismiss-new-order-modal

Juanpe committed:
Escaping discard closure of the discard changes action sheet

Juanpe commented:
Hi again @itsmeichigo, I made all the suggested changes, I hope I didn't miss any 😅

As for unit tests, I thought about adding some unit tests to cover when the `shouldShowCancelButton` is true/false based on the value of the feature flag, but I think it's a temporary thing until the feature flag is definitely enabled. So, IMO, it's not worth it. WDYT?

Thanks again for the feedback!

Juanpe committed:
Define the `FeatureFlagService` as a dependency of `NewOrderViewModel`

Juanpe committed:
Merge branch 'trunk' into issue/6500-add-button-to-dismiss-new-order-modal

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -66,16 +66,48 @@ extension NewOrderHostingController: UIAdaptivePresentationControllerDelegate {
+            return
+        }
+
+        self.dismiss(animated: true)
```
Nit: we can remove the `self.` in this following our coding convention.

Juanpe replied:
Done ✅ Thanks!

itsmeichigo reviewed (commented):
Thanks for the updates! Overall I think it's all good now, however, I'm concerned about the consistency in the check for changes among different modals, so I've asked in the internal discussion here: p1649383906341789-slack-C6H8C3G23.

Let's wait for the team's opinions before merging this, just to be sure.

> As for unit tests, I thought about adding some unit tests to cover when the shouldShowCancelButton is true/false based on the value of the feature flag, but I think it's a temporary thing until the feature flag is definitely enabled. So, IMO, it's not worth it. WDYT?

Agreed - let's skip the unit test for now.

Juanpe committed:
Remove unnecessary `self.`

Juanpe committed:
Merge branch 'trunk' into issue/6500-add-button-to-dismiss-new-order-modal

Juanpe committed:
Merge branch 'trunk' into issue/6500-add-button-to-dismiss-new-order-modal

Juanpe committed:
Remove check unsaved changes when tapping Close button

Juanpe commented:
> Thanks for the updates! Overall I think it's all good now, however, I'm concerned about the consistency in the check for changes among different modals, so I've asked in the internal discussion here: p1649383906341789-slack-C6H8C3G23.
> 
> Let's wait for the team's opinions before merging this, just to be sure.
> 

Based on the internal discussion, we don't have to check for changes when tapping the Close button on the New Order screen. So, I've removed the check 👍🏼 


**📺 Video** showing we check the unsaved changes when swiping down but not when tapping on the Close button..

https://user-images.githubusercontent.com/1409041/162420998-da22b369-2c63-4af6-99b4-c7a304a4dd39.mp4

Juanpe committed:
Update UI tests doc

itsmeichigo reviewed (approved):
LGTM :shipit:
